Run > Files > External Data fixes

bodom0015 opened the pull request:
This is a rather large / inter-dependent PR, but sadly these are very related - to the point that it would take a lot longer to do/review them as separate PRs.

I apologize in advance for the complexity here.

## Related Issues
* ~Depends on #359~
* ~Depends on https://github.com/whole-tale/girder_wholetale/pull/218~
* ~Depends on https://github.com/whole-tale/girder_wholetale/pull/213~
* ~Depends on https://github.com/whole-tale/gwvolman/pull/45~
* ~Depends on https://github.com/whole-tale/girder_wt_data_manager/pull/36~
* Fixes #355 - External Data nav on Run > Files tab should now be reading from `tale.dataSets` instead of `/dm/session` or `tale.involatileData`
* Fixes #354 - Left pane of Select Data modal should now list from the `WholeTale Catalog`, instead of user's own registered datasets
* Fixes #350 - added a hacky `resync()` function that recycles existing logic
* Fixes #352 - `allSelectedItems` should now be built up from the `dataSet` object on the tale
* Fixes #353 - `this.session.save()` was being used, so I adjusted this to save the selected dataSets to `/tale/:id` instead, and the duplicate modal no longer launches
* If I can figure it out quickly enough, I may attempt to add #354 here as well (since it will also depend on all of the above)

## Changes
* Consolidated duplicate "Select Data" modal logic - moved this down from the `run` component into the `tale-tab-files` component
* Changed remaining references in the `run` / `tale-tab-files` components from `tale.involatileData` to instead use `tale.dataSet`
* Changed `this.session.save()` to instead call PUT `/tale/:id` to save selected data from the "Select Data" modal

## How to Test
Assuming all related/dependent PRs have been merged:
1. Use [the deploy-dev "dev" branch](https://github.com/whole-tale/deploy-dev/tree/dev) to map in and run all of the dependent branches together (see above)
2. Login to Whole Tale Dashboard
3. If you haven't already, register a Dataset (e.g. `doi:10.18739/A29G5GD0V`)
    * NOTE: You may need to wait for registration to complete
4. Compose a new Tale if you haven't already
    * NOTE: Tale version has been updated to 4
    * If you choose an older Tale, please ensure that it has migrated properly to version 4 (e.g. `format` should be `4`, and you should now see ` dataSet` instead of `involatileData`)
5. Launch the Tale, navigate to Run > Files, and click "External Data"
    * You should not yet see any datasets listed here
6. At the top-right of the panel, click the (+) button
    * You should be presented with the "Select Data" modal
    * On the left, you should see all datasets registered by all users (#354) - NOTE: This no longer needs to match the listed output from Manage > Data, and should instead be the union of all users' registered datasets
    * The right side should match the External Data nav of the Run > Files tab - e.g. you should not yet see any datasets listed here
7. Add the dataset you just registered by selecting it on the left and clicking "Add Selected"
    * You should see the registered dataset appear on the right side
8. Click "Select" at the bottom-right of the modal window
    * The modal window should dismiss itself
    * You should still be at the Run > Files view with the External Data nav selected
    * You should see the selected data has resynced and now appears in the file list (#353)
    * You should no longer see a duplicate "Select Data" modal after dismissing the first one (#353)
9. Using SwaggerUI, examine the Tale object you just created
    * You should see the `dataSet` field has been populated with the id of the dataSet you just added (#350)
10. Click the (+) button to open the "Select Data" modal again
    * You should see the data you just added now appears on the right side, which should still match your "External Data" file list from the Run > Files view (#352)

Xarthisius commented:
@bodom0015 step 9. fails for me. I created a new Tale without external data initially. When I use the modal everything looks fine, but `dataSet` is not updated in the Tale object. When I refresh the page it's also gone from the UI. I see some errors in the console:

1. When switching to 'external data' tab in a running Tale or closing modal via "select" button:
```
rsvp.js:27 Uncaught ReferenceError: session is not defined
    at taleDatasetContents.then._taleDatasetContents (component.js:263)
    at tryCatcher (rsvp.js:200)
    at invokeCallback (rsvp.js:372)
    at publish (rsvp.js:358)
    at rsvp.js:14
    at invoke (backburner.js:205)
    at Queue.flush (backburner.js:125)
    at DeferredActionQueues.flush (backburner.js:278)
    at Backburner.end (backburner.js:410)
    at Backburner._runExpiredTimers (backburner.js:831)
```
2. 400 on PUT /tale (which explains lack of update):

```
{
  message: "Invalid JSON object for parameter tale: None is not of type 'string'",
  type: "rest"
}
```
is this the issue with `description`?

bodom0015 commented:
> @bodom0015 step 9. fails for me. I created a new Tale without external data initially. When I use the modal everything looks fine, but `dataSet` is not updated in the Tale object. When I refresh the page it's also gone from the UI. I see some errors in the console:
> 
> 1.)When switching to 'external data' tab in a running Tale or closing modal via "select" button:
> 
> ```
> rsvp.js:27 Uncaught ReferenceError: session is not defined
>     at taleDatasetContents.then._taleDatasetContents (component.js:263)
>     at tryCatcher (rsvp.js:200)
>     at invokeCallback (rsvp.js:372)
>     at publish (rsvp.js:358)
>     at rsvp.js:14
>     at invoke (backburner.js:205)
>     at Queue.flush (backburner.js:125)
>     at DeferredActionQueues.flush (backburner.js:278)
>     at Backburner.end (backburner.js:410)
>     at Backburner._runExpiredTimers (backburner.js:831)
> ```
Ah, good catch! I had tried to clean up some commented things and accidentally left one uncommented. I'll get rid of that

> 2.) 400 on PUT /tale (which explains lack of update):
> 
> ```
> {
>   message: "Invalid JSON object for parameter tale: None is not of type 'string'",
>   type: "rest"
> }
> ```
> is this the issue with `description`?

Yes this is the same issue and error message that is described in https://github.com/whole-tale/girder_wholetale/issues/215

Xarthisius commented:
:tada: 
![pr](https://user-images.githubusercontent.com/352673/51629171-80a6e280-1f0c-11e9-90df-cfb3731f9441.gif)

bodom0015 commented:
Updated this with a quick fix for #354 to unblock the work needed for #357.

Minor changes:
1. The contents of the Left side of the "Select Data Modal" from Run > Files > External Data should now list the "WholeTale Catalog", instead of the only the users' registered datasets.
2. I also added a quick / hacky fix for an annoying problem I've been encountering: very rarely, the Select Data Modal was refusing to close when I would click "Select". I now explicitly call the `closeSelectDataModal` method to ensure that it is hidden properly when the user confirms the modal.

Xarthisius commented:
Great effort! Thanks @bodom0015 !
